### PR TITLE
osd: load latest cephx key during osd pod init

### PR DESCRIPTION
**Pull request by BlaineEXE**
Create a new OSD pod init container that loads the latest CephX key for
the OSD. The init container must be run after OSD activation, due to
keyfile reset behavior of ceph-bluestore-tool, used during activation.

The primary purpose of this work is to allow OSDs to start and run
correctly after key rotation, whether done manually by users or
automatically by future Rook work.

This work makes redundant prior code changes done in https://github.com/rook/rook/pull/14826. The CI test
added in the prior work is kept to ensure no regression.

Manual testing done:

- Host-based OSDs
  - [x] manual OSD key rotation - OSDs work after activate init container added
  - [x] manual bootstrap-osd key rotation - provision more OSDs after rotation works w/o changes
- PVC-based OSDs
  - [x] manual OSD key rotation
  - [x] manual bootstrap-osd key rotation

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 1]
```diff
@@ -111,38 +111,10 @@ KEYRING_FILE="$OSD_DATA_DIR"/keyring
...
-        rm --recursive --force "$TMP_DIR"
-    else
-        # create new keyring if no keyring file found
-        ceph -n client.admin auth get-or-create osd."$OSD_ID" mon 'allow profile osd' mgr 'allow profile osd' osd 'allow *' -k /etc/ceph/admin-keyring-store/keyring
-    fi
-fi
+# always store the latest keyring file, in case key generation during OSD prepare failed, or in case of OSD key rotation
+ceph -n client.admin auth get-or-create osd."$OSD_ID" \
+  mon 'allow profile osd' mgr 'allow profile osd' osd 'allow *' \
+  -k /etc/ceph/admin-keyring-store/keyring > "$KEYRING_FILE"
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/spec.go:117]: So this is really strange. I looked into `ceph-volume raw activate`, and it doesn't seem to be touching the keyring at all. It does call `ceph-bluestore-tool` which I suspect is grabbing the key from the bluestore contents. In my testing, refreshing the key at the end of 'activate' resolves things, at least for raw OSDs on hosts.

I still have to test PVC OSDs, and I suspect the OSD bootstrap keyring may have issues given its current implementation.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 2]
```diff
@@ -218,6 +190,12 @@ sys.exit('no disk found with OSD ID $OSD_ID')
 	# ceph-volume raw mode only supports bluestore so we don't need to pass a store flag
 	ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
 fi
+
+# something in the 'ceph-volume activate' command resets the key to a prior version, so refresh the
+# keyring file again at the end of activate init
+ceph -n client.admin auth get-or-create osd."$OSD_ID" \
+  mon 'allow profile osd' mgr 'allow profile osd' osd 'allow *' \
+  -k /etc/ceph/admin-keyring-store/keyring > "$KEYRING_FILE"
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/spec.go]: So this is really strange. I looked into `ceph-volume raw activate`, and it doesn't seem to be touching the keyring at all. It does call `ceph-bluestore-tool` which I suspect is grabbing the key from the bluestore contents. In my testing, refreshing the key at the end of 'activate' resolves things, at least for raw OSDs on hosts.

I still have to test PVC OSDs, and I suspect the OSD bootstrap keyring may have issues given its current implementation.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 3]
```diff
@@ -1430,6 +1408,66 @@ func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps
...
 }
 
+// init container to get (or create, if needed) the most up-to-date keyring for the given OSD and
+// update the OSD config directory keyring file. this allows OSD cephx keys to be rotated without
+// causing OSD pod crash loops.
+//
+// this must come after OSD activate init container because 'ceph-bluestore-tool' (and consequently,
+// 'ceph-volume activate') pulls the keyring from the OSD disk's bluestore data and overwrites the
+// config dir's keyring file, even when the on-disk key is not the most up-to-date.
+func (c *Cluster) getCephxKeyUpdateInitContainer(osdID string, osdProps osdProperties) (v1.Volume, v1.Container) {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/spec.go:1414]: The cephx key update doesn't **have** to be part of the activate container.

Instead of modifying the 3 current activate container implementations (including the deprecated LVM PVC OSDs), it seems like the least code churn if we create a new init container to handle CephX key updates. Testing that out now here.

**Comment by BlaineEXE** [comment]
FYI @prazumovsky. I wanted to let you know that the implementation you wrote for loading the correct OSD cephx keys is being modified slightly here to support cephx key rotation. The CI test you wrote for your changes is still present and should be verifying that there is no regression. If you'd like, feel free to review this PR to double check. Thanks!

I found loading and storing the latest `ceph auth` key after OSD activation was the simplest way to support rotation as well as keep the CI test you wrote healthy.

OSD activation calls `ceph-bluestore-tool`. In my testing, the tool is pulling the OSD's last-working keyring from the OSD bluestore on-disk metadata and storing it in the keyring file, meaning that activation has to happen before key update. I didn't see any changes in OSD init or runtime behavior by updating the key before activation.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 4]
```diff
@@ -111,39 +111,6 @@ KEYRING_FILE="$OSD_DATA_DIR"/keyring
...
-        cat "$TMP_DIR"/keyring
-        ceph -n client.admin auth import -i "$TMP_DIR"/keyring -k /etc/ceph/admin-keyring-store/keyring
-
-        rm --recursive --force "$TMP_DIR"
-    else
-        # create new keyring if no keyring file found
-        ceph -n client.admin auth get-or-create osd."$OSD_ID" mon 'allow profile osd' mgr 'allow profile osd' osd 'allow *' -k /etc/ceph/admin-keyring-store/keyring
-    fi
-fi
-
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/spec.go:146]: We don't **have** to remove this for key rotation to work. However, by removing it, the CI test associated with the added code still passes, so I am quite sure it's unnecessary at this point, unless prazumovsky believes it is still necessary for his corner case.

Or if you think it's worth it to keep "just in case," let me know, and I can add it back in

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 5]
```diff
@@ -651,6 +618,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd *OSDInfo, provision
 		initContainers = append(initContainers, c.getExpandInitContainer(osdProps, c.spec.DataDirHostPath, c.clusterInfo.Namespace, osdID, *osd))
 	}
 
+	// key update init container must go after activate
+	adminKeyVol, cephxContainer := c.getCephxKeyUpdateInitContainer(osdID, osdProps)
+	if !volumeExistsWithName(volumes, adminKeyVol.Name) {
+		volumes = append(volumes, adminKeyVol)
+	}
+	initContainers = append(initContainers, cephxContainer)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go:626]: Does this init container not need to be before the activate container? I wonder if the activate could fail if this does not come first.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 6]
```diff
@@ -1430,6 +1404,86 @@ func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps
...
+	} else {
+		// assumption: activate OSD volume will already be added to pod spec for non-PVC OSDs at the
+		// correct location for mounting to /var/lib/ceph/osd/ceph-<id>
+		volMounts = append(volMounts,
+			v1.VolumeMount{Name: activateOSDVolumeName, MountPath: osdConfigMountPath},
+		)
+	}
+
+	keyUpdateScript := `
+set -o errexit
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/spec.go:1437]: This script looks much shorter than where it was removed from the activate init container. Is the difference some legacy osd init? I don't fully recall the context around that script.

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 7]
```diff
@@ -1430,6 +1404,86 @@ func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps
 	}
 }
 
+// init container to get (or create, if needed) the most up-to-date keyring for the given OSD and
+// update the OSD config directory keyring file. this allows OSD cephx keys to be rotated without
+// causing OSD pod crash loops.
+//
+// this must come after OSD activate init container because 'ceph-bluestore-tool' (and consequently,
+// 'ceph-volume activate') pulls the keyring from the OSD disk's bluestore data and overwrites the
+// config dir's keyring file, even when the on-disk key is not the most up-to-date.
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/spec.go:1413]: Note

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 5 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/spec.go:626]: It must come after the activate container. I tried to capture that in code comments [here](https://github.com/rook/rook/pull/16003/files#r2153165796), but let me know if that isn't clear.

Also, I did test this thoroughly for both host-based and PVC-based (non-LVM) OSDs. If this comes before activate, activate overwrites the key with the last-running key, which is not the most up-to-date key, causing OSDs to fail startup. When this is after, OSDs get the key updates and run without errors.

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 6 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/spec.go:1437]: I would say that it just makes better use of Ceph's tooling. There really isn't a need for if-notexist-then-create-else-get when `ceph auth get-or-create` exists

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 5 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/spec.go:626]: I missed that comment, that is clear now, thanks

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 8]
```diff
@@ -1430,6 +1404,86 @@ func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps
 	}
 }
 
+// init container to get (or create, if needed) the most up-to-date keyring for the given OSD and
+// update the OSD config directory keyring file. this allows OSD cephx keys to be rotated without
```
Comment (sp98) [inline pkg/operator/ceph/cluster/osd/spec.go:1408]:
```suggestion
// update the OSD config directory keyring file. This allows OSD cephx keys to be rotated without
```

File: pkg/operator/ceph/cluster/osd/spec.go [hunk 9]
```diff
@@ -1430,6 +1404,86 @@ func (c *Cluster) getEncryptedStatusPVCInitContainer(mountPath string, osdProps
...
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o xtrace
+
+OSD_ID="` + osdID + `"
+KEYRING_FILE=/var/lib/ceph/osd/ceph-"${OSD_ID}"/keyring
+
+ls /etc/ceph/admin-keyring-store
+cat /etc/ceph/admin-keyring-store/keyring
```
Comment (sp98) [inline pkg/operator/ceph/cluster/osd/spec.go]: is it ok to log the keyring?

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 8 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/spec.go:1408]: I'm not as concerned about grammar in code comments, thanks

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 9 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/spec.go]: GOOD CATCH! This was just for my own debugging. I'll strip that

File: pkg/operator/ceph/cluster/osd/spec.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/osd/spec.go:146]: I wonder if in that case the activate was failing without that fix. It might be safer to leave it in for now. @prazumovsky thoughts?

Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/osd/spec.go:146]: Update from huddle discussion:

I'm not able to recreate the corner case issue that was reported associated with this fix, so I can't know for certain that things are fixed. But I do believe that the changes here still address the issue, though via different means.

The code being removed here attempts to load the keyring from the OSD info in /var/lib/rook. If that file isn't found, then it creates a new one.

The changes here flip the paradigm around. It stops treating the /var/lib/rook key as being the source of truth (that makes key rotation unsupportable afaict). Instead, we treat the mon cluster as source of truth, and update the /var/lib/rook key based on that.

The new pattern mirrors what's done for most of the other daemons where we run `auth get-or-create` and then put that result into the corresponding secret, which in turn appears inside the daemon pod. Here for OSDs, the result just goes stright to disk instead, with no secret.

Because we have a reasonably high confidence that the change should work in the corner case, we will continue for now. If we hear back about the corner case returning, I/we will follow up to add it back.
